### DISCUSSION TITLE
Add ability to update Time Of Use settings

### DISCUSSION THREAD

**Discussion by crudler** [StephanJoubert/home_assistant_solarman#347] [Ideas]
Similar to how the Deye app allows remote control functions, it would be useful to allow for automation based on rules

ie. when loadshedding reaches stage 4, or solar forecasting is low, change to grid charge

**Comment by NikolajChristensen** [comment]
This would be a great addition. How about it? :)

**Comment by PaulOckenden** [comment]
I’d really love this too. It would take this integration to the next level. I’m using a Sofar Solar ME3000SP and would love to be able to flip between TOU and auto modes on both of my inverters.

**Comment by NikolajChristensen** [comment]
You can actually with the new ModBus service.

**Reply by PaulOckenden** [reply]
Really? Have you tried? Does it work? Looks like there's lots of config needed, so any pointers would be appreciated!

**Reply by NikolajChristensen** [reply]
Yes. I have used it to change the maximum battery discharge rate. You can see the codes here:

https://github.com/klatremis/esphome-for-deye/blob/main/deye12k%20esphome%20config?fbclid=IwAR2V-U2pBxR3rO7RxwDmwUP3pvFaxS6Yttc9gTXEYjfSD2InipA6Nj7dR3A

In HA it looks something like the attached. 
![img_1_1697301660879](https://github.com/StephanJoubert/home_assistant_solarman/assets/17816446/96cbb500-d3b9-44b3-83a1-5cc3ad351bb4)

**Reply by bazzar1206** [reply]
Hi. Does this require another device, the esp thing, or can it just be used via the logger of the inverter? That link you have provided does nothing but confuse me! Thanks.

**Reply by NikolajChristensen** [reply]
This can be done with the Solarman logger only. Look at the screenshot. That screenshot changes address 145. In the link you can find the addresses for other functionality. It's just a coincidence that the document is at a github related to esphome. You can ignore that.

**Comment by NikolajChristensen** [comment]
This is an example from the document:

platform: modbus_controller
    use_write_multiple: true
    modbus_controller_id: ${modbus_controller_id}
    name: ${device_type}_Solar_sell
    register_type: holding
    address: 145
    bitmask: 1
    entity_category: config
    icon: "mdi:toggle-switch"

So address 145 is solar sell on the inverter.

**Reply by bazzar1206** [reply]
Thank you for replying. I sort of get you. I am so far out my depth with this. Trying to understand but may just leave it. But really appreciate you taking time to reply.

**Reply by NikolajChristensen** [reply]
Try reproducing the screenshot I posted and see how solar cell toggles on your inverter. Maybe that makes it click :). Good luck.

**Comment by ifilgud** [comment]
It would be wonderful to have this functionality via Solarman integration. I have an electric car and to avoid unnecesary charge cycles of battery it would be nice to configure minimum battery % in the time of use config. Today I've learned how to change it via Deye Cloud app, so making an automation would be nice (dreaming is free :D)

**Comment by RaoulSargent** [comment]
I too would love to use the HA solarman integration with an automation to change the TOU and force charge/discharge times... using with Octopus Agile tariff in UK.
My inverter is: Solis Hybrid model RHI-5K-48ES-5G

**Reply by bristolsam1** [reply]
Hi RaoulSargent, have you tried writing to registry on Solis inverter. does it work?

**Comment by RaoulSargent** [comment]
@bristolsam1 Yes indeed I have.

I have several automations in daily use:  [I am on Octopus Flux Export/Import]
- if batteries > 80% @ 18:00 then force export 18:00 - 19:00 (export surplus during max export rate)
- if batteries < 55% @ 14:00 then force import 14:00 - 16:00 (top up batteries in mid import rate before max import rate starts)
- 02:00 - 05:00 force charge batteries during cheapest import rate (qty kWh imported varies according to month of year)

I am not currently doing anything with predicting 'tomorrows' weather.
For the moment I decided that it was enough to always ensure things like the batteries always had sufficient kWh stored to avoid needing to import during the max import rate 16:00-19:00 each day.

Also that I always import at least 3 kWh during the 02:00-05:00 (cheapest import rate) regardless of battery SOC.
On good sunny days this does mean that I import eg 3x kWh that I didn't need to do, which means the batteries reach 100% SOC sooner than they would have done and then I start exporting sooner than I would have done.
The worst scenario is that I export at the same rate I imported, but the benefit is that if there is no sun then I run off battery during the early morning hours before sun rises and thus benefit from the cheaper import rate.

I have been doing this since November 2023 and typically been saving approx £1/day just by buying kWh in a cheaper rate when the batteries are too low for the next more expensive rate, and that is not including solar savings.  My savings per day are much higher when including solar of course.

This little chart shows "Average Savings Per Day" calculated over different timeframes.
It is based on what the cost of electric would have been at the time it was consumed compared to the actual imported cost for the day and includes the value of any export.

<img width="304" alt="image" src="https://github.com/user-attachments/assets/4ad97666-252e-48aa-bd4c-1c70e4cad5ad">



**What am I using to do this ?**

18x 400Wh panels in 3x strings of 6-South / 7-South / 5-West panels (2x Solis inverters)
1 inverter is grid connected and has 4x US3000C Pylontech batteries.

Home Assistant with the Solarman integration.
To control the Solis Inverter and change its TOU settings (by writing to the registers) I am using a python script and HA integration.
'pyscript 1.6.1'

I have another automation that runs twice every day to set the clock of the inverter.  I found my inverter clock/time drifts by a few seconds a week and hence needs to be keep in sync else would not match the Octopus time slots.  [My inverter is firewalled so cannot connect to the internet so cannot update its own time]

**WARNING: Use at your own risk... I do not know if the register values I use are the same for your inverter!!**

This is my "Force Discharge" flow if the batteries SOC > 80% at 18:00 each day...

<img width="485" alt="image" src="https://github.com/user-attachments/assets/eab8e057-7b9d-4732-821f-51712346b090">

**WARNING: Use at your own risk... I do not know if the register values I use are the same for your inverter!!**

Example:

<img width="1014" alt="image" src="https://github.com/user-attachments/assets/4070bd36-9323-4a65-9a8c-330b44bb9324">

**WARNING: Use at your own risk... I do not know if the register values I use are the same for your inverter!!**

I have another automation that runs every day to always ensure my 'defaults' are back in place before the 02:00-05:00 slot starts.

Please let me know if you need any further info.

**What next...**

I now have 3+ years worth of Solis data in a MariaDB database, so I intend to use that historical data, combine with weather prediction and solar forecasts to dynamically adjust how much I force import/export each day... however I feel this is more of an academic exercise rather than going to deliver much benefit over what I already achieve.

The other thing I am planning to do is (possibly) move to Octopus Agile.
But before I do that I want to write a 'model' to compare my actual usage/savings with Flux vs Agile.
(This would be more detailed than the Octopus online compare tool that works on daily average rates).

**Reply by bristolsam1** [reply]
Thanks for the detail reply, Yes I managed to write data to the Solis inverter and I will use your automations to change timing of discharge and charge cycles. I am currently on Intelligent Octopus GO as I have an EV and will use this for the winter period and may change to Flux in summer time, and need to compare with Agile. If you find a good model for the comparison please let me know.
